src/python/pants/backend/go/build.py

```diff
@@ -0,0 +1,354 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+import logging
+import os
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import Dict, List, Tuple
+
+from pants.backend.go.distribution import GoLangDistribution
+from pants.backend.go.target_types import GoBinaryMainAddress, GoBinaryName, GoImportPath, GoSources
+from pants.build_graph.address import Address, AddressInput
+from pants.core.goals.package import (
+    BuiltPackage,
+    BuiltPackageArtifact,
+    OutputPathField,
+    PackageFieldSet,
+)
+from pants.core.util_rules.external_tool import DownloadedExternalTool, ExternalToolRequest
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import AddPrefix, CreateDigest, Digest, FileContent, MergeDigests, Snapshot
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.platform import Platform
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import collect_rules, goal_rule, rule
+from pants.engine.target import (
+    FieldSet,
+    Targets,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+    WrappedTarget,
+)
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+_logger = logging.getLogger(__name__)
+
+
+class GoBuildSubsystem(GoalSubsystem):
+    name = "go-build"
```

Comment:
It's unclear what the future of this goal should be, e.g. `compile`. Relates to work Patrick is doing with Java. But this is totally fine as an incremental checkpoint.